packages/schemas/src/text/uiRender.ts

```diff
@@ -51,28 +33,120 @@ export const uiRender = async (arg: UIRenderProps<TextSchema>) => {
+  const usePlaceholder = isEditable(mode, schema) && placeholder && !value;
+
+  const textBlock = await buildStyledTextContainer(arg, usePlaceholder ? placeholder : value);
+
+  if (!isEditable(mode, schema)) {
+    // Read-only mode
+    textBlock.innerHTML = value
+        .split('')
+        .map(
+            (l: string, i: number) =>
+                `<span style="letter-spacing:${
+                    String(value).length === i + 1 ? 0 : 'inherit'
+                };">${l}</span>`
+        )
+        .join('');
+    return;
+  }
+
+  makeElementPlainTextContentEditable(textBlock);
+  textBlock.tabIndex = tabIndex || 0;
+  textBlock.innerText = value;
+  textBlock.addEventListener('blur', (e: Event) => {
+    onChange && onChange({ key: 'content', value: (e.target as HTMLDivElement).innerText });
+    stopEditing && stopEditing();
+  });
+
+  if (schema.dynamicFontSize) {
+    let dynamicFontSize: undefined | number = undefined;
+    const font = options?.font || getDefaultFont();
+    const fontKitFont = await getFontKitFont(schema.fontName, font, _cache);
+
+    textBlock.addEventListener('keyup', () => {
+      setTimeout(() => {
+        void (async () => {
+          if (!textBlock.textContent) return;
+          dynamicFontSize = await calculateDynamicFontSize({
+            textSchema: schema,
+            font,
+            value: textBlock.textContent,
+            startingFontSize: dynamicFontSize,
+            _cache,
+          });
+          textBlock.style.fontSize = `${dynamicFontSize}pt`;
+
+          const { topAdj: newTopAdj, bottomAdj: newBottomAdj } =
+            getBrowserVerticalFontAdjustments(
+              fontKitFont,
+              dynamicFontSize ?? schema.fontSize ?? DEFAULT_FONT_SIZE,
+              schema.lineHeight ?? DEFAULT_LINE_HEIGHT,
+              schema.verticalAlignment ?? DEFAULT_VERTICAL_ALIGNMENT
+            );
+          textBlock.style.paddingTop = `${newTopAdj}px`;
+          textBlock.style.marginBottom = `${newBottomAdj}px`;
+        })();
+      }, 0);
+    });
+  }
+
+  if (usePlaceholder) {
+    textBlock.style.color = PLACEHOLDER_FONT_COLOR;
+    textBlock.addEventListener('focus', () => {
+      if (textBlock.innerText === placeholder) {
+        textBlock.innerText = '';
+        textBlock.style.color = schema.fontColor ?? DEFAULT_FONT_COLOR;
+      }
+    });
+  }
+
+  if (mode === 'designer') {
+    setTimeout(() => {
+      textBlock.focus();
+      // Set the focus to the end of the editable element when you focus, as we would for a textarea
+      const selection = window.getSelection();
+      const range = document.createRange();
+      if (selection && range) {
+        range.selectNodeContents(textBlock);
+        range.collapse(false); // Collapse range to the end
+        selection?.removeAllRanges();
+        selection?.addRange(range);
+      }
+    });
+  }
+};
+
+export const buildStyledTextContainer = async (arg: UIRenderProps<TextSchema>, value: string) => {
```

Comment:
I have refactored this file into 2 functions that can be shared by the multi-variable text schema - `buildStyledTextContainer` and `makeElementPlainTextContentEditable`.

I think this makes this file a bit simpler and easy to read too.